In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Setting the path of data
dir_data = './Day2data/'
train = os.path.join(dir_data, 'application_train.csv')
train_data = pd.read_csv(train)
train_data.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

              ...              FLAG_DOCUMENT_18 FLAG_DOCUMENT_19  \
0             ...                             0                0   
1             ...                             0                0   
2             ...                             0                0   
3             ...                             0                0   
4             ...                             0                0   

  FLAG_DOCUMENT_20 FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR  \
0                0                0                        0.0   
1                0                0                        0.0   
2                0                0                        0.0   
3                0                0                        NaN   
4                0                0                        0.0   

  AMT_REQ_CREDIT_BUREAU_DAY  AMT_REQ_CREDIT_BUREAU_WEEK  \
0                       0.0                         0.0   
1                       0.0                         0.0   
2                       0.0                         0.0   
3                       NaN                         NaN   
4                       0.0                         0.0   

   AMT_REQ_CREDIT_BUREAU_MON  AMT_REQ_CREDIT_BUREAU_QRT  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   
3                        NaN                        NaN   
4                        0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_YEAR  
0                         1.0  
1                         0.0  
2                         0.0  
3                         NaN  
4                         0.0  

[5 rows x 122 columns]

In [3]:
# in order to get the q0-q100, first i generate a np.array that contains values from 0 - 100.
all_num = np.arange(0,101,1)
q_all = [np.percentile(train_data[~train_data['AMT_ANNUITY'].isnull()]['AMT_ANNUITY'], q = i) for i in all_num]
print("The q0 - q100 of \''AMT_ANNUITY\'' is as followed: \n")
print(q_all)

a = pd.DataFrame({'q': all_num, 'value': q_all})

The q0 - q100 of ''AMT_ANNUITY'' is as followed: 

[1615.5, 6182.910000000001, 6750.0, 7875.0, 8703.0, 9000.0, 9000.0, 9553.5, 10125.0, 10503.0, 11074.5, 11430.0, 11970.0, 12375.0, 12838.5, 13302.0, 13500.0, 13500.0, 13896.0, 14350.5, 14701.5, 15124.5, 15583.5, 15970.5, 16209.0, 16524.0, 16852.5, 17109.0, 17487.0, 17806.5, 18189.0, 18643.5, 19102.5, 19417.5, 19836.0, 20151.0, 20421.0, 20853.0, 21186.0, 21609.0, 21865.5, 22018.5, 22342.5, 22527.0, 22972.5, 23346.0, 23719.5, 23931.0, 24259.5, 24583.589999999953, 24903.0, 25240.5, 25537.5, 25960.5, 26217.0, 26316.0, 26640.0, 26860.5, 27189.0, 27558.0, 28062.0, 28480.5, 28917.0, 29340.0, 29830.5, 30078.0, 30483.0, 30897.0, 31275.0, 31630.5, 32004.0, 32458.5, 32895.0, 33376.5, 33984.0, 34596.0, 35345.16000000005, 35806.5, 36328.5, 36747.0, 37516.5, 37948.5, 38556.0, 39456.0, 40135.5, 40806.0, 41845.5, 42790.5, 43735.5, 44991.0, 45954.0, 47254.5, 48465.0, 49878.0, 51745.5, 53325.0, 55624.5, 58482.0, 62964.0, 70006.5, 258025.5]


In [4]:
# try to replace the NAs by q50 
print("Before replace NAs, numbers of row that AMT_ANNUITY is NAs: %i" % sum(train_data['AMT_ANNUITY'].isnull()))

# q50 is the value number 50 in a DataFrame that just generated by myself
q_50 = a["value"][50]
train_data.loc[train_data['AMT_ANNUITY'].isnull(),'AMT_ANNUITY'] = q_50

print("After replace NAs, numbers of row that AMT_ANNUITY is NAs: %i" % sum(train_data['AMT_ANNUITY'].isnull()))

Before replace NAs, numbers of row that AMT_ANNUITY is NAs: 12
After replace NAs, numbers of row that AMT_ANNUITY is NAs: 0


In [5]:
# in order to normalize this column, i print out the data description before the normalization
print("== Original data range ==")
print(train_data['AMT_ANNUITY'].describe())

== Original data range ==
count    307511.000000
mean      27108.487841
std       14493.461065
min        1615.500000
25%       16524.000000
50%       24903.000000
75%       34596.000000
max      258025.500000
Name: AMT_ANNUITY, dtype: float64


In [6]:
# define a normalization function as down below
def normalize_value(x):
    
    x = 2*((x-x.min())/(x.max()-x.min()))-1

    return x

In [7]:
# normalize the "AMT_ANNUITY" column by normalization function called "normalize_value".
train_data['AMT_ANNUITY_NORMALIZED'] = normalize_value(train_data.AMT_ANNUITY)

# after normalization, I print out the data description
print("== Normalized data range ==")
train_data['AMT_ANNUITY_NORMALIZED'].describe()

== Normalized data range ==


count    307511.000000
mean         -0.801154
std           0.113049
min          -1.000000
25%          -0.883714
50%          -0.818357
75%          -0.742752
max           1.000000
Name: AMT_ANNUITY_NORMALIZED, dtype: float64

as you can see, the maximun value is 1 and the minimun is -1, which confirm that the normalization function works!!!

In [8]:
# try to get the most frequently value ( mode ) in "AMT_GOODS_PRICE" column
# and replace the NAs by the mode of "AMT_GOODS_PRICE" column


# so i calculate the number of NAs in this column and print out the data description of the column
print("Before replace NAs, numbers of row that AMT_GOODS_PRICE is NAs: %i" % sum(train_data['AMT_GOODS_PRICE'].isnull()))
print(train_data['AMT_GOODS_PRICE'].describe())

Before replace NAs, numbers of row that AMT_GOODS_PRICE is NAs: 278
count    3.072330e+05
mean     5.383962e+05
std      3.694465e+05
min      4.050000e+04
25%      2.385000e+05
50%      4.500000e+05
75%      6.795000e+05
max      4.050000e+06
Name: AMT_GOODS_PRICE, dtype: float64


In [9]:
# as you can see, there are 278 NAs in this column

# get the mode of this column by using the code down below
from statistics import mode
value_most = mode(train_data['AMT_GOODS_PRICE'])
print(value_most)

450000.0


In [10]:
# The mode of this column is 450000.0 which means that 450000.0 is the most frequently value in this column
# and we can calculate the number of value 450000.0 by the code down below
count = 0
for i in range(307233):
    if train_data['AMT_GOODS_PRICE'][i] == 450000.0:
        count += 1
print(count)

26003


In [11]:
# replace the NAs by the mode of this column
train_data.loc[train_data['AMT_GOODS_PRICE'].isnull(),'AMT_GOODS_PRICE'] = value_most

In [12]:
print("After replace NAs, numbers of row that AMT_GOODS_PRICE is NAs: %i" % sum(train_data['AMT_GOODS_PRICE'].isnull()))
count = 0
for i in range(307233):
    if train_data['AMT_GOODS_PRICE'][i] == 450000.0:
        count += 1
print(count)

After replace NAs, numbers of row that AMT_GOODS_PRICE is NAs: 0
26281


As you can see, the number of value of 450000.0 is 26281. 
Why? because we replace the NAs by the mode.
And the number of NAs in this column before the replacement is 278
and after the replacement, the number of NAs becomes to 0 and the number of value of 450000.0 becomes to 26281 (from 26003)
The difference of 26281 and 26003 is exactly the number of the NAs before the replacement.
So, the replacement is successfully !!!